rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -116,6 +134,47 @@ to compute approximate heavy hitters over this in-memory representation. We can
 now share this implementation inside both TFDV and TFT for their top-K feature
 value computation.
 
+# Requirements
+
+The in-memory representation should:
+
+*   Be columnar.
+
+    A columnar in-memory representation works better than a row based on under
+    typical workload of TFX libraries:
+    +   Compute statistics over a column.
+    +   Feed a batch of rows to TensorFlow.
```

Comment:
Why is this easier when columnar versus row-based? This seems to be an argument for row-based.

Reply:
- Batching rows into batch-aligned tensors is in-efficient.
- computing (aggregating) statistics row-by-row is in-efficient.

The inefficiency came from python loops / function calls, and will be from having to read a columnar storage in a row-based manner if the storage is columnar.

Reply:
Karmel, currently when you feed multi-column data into TF the most efficient way to do so is to have one tensor per column with one row in the tensor per row in the data. I think this is what this is alluding to?

That said I still don't know why the in-memory representation here cannot just be an eager StructuredTensor

Reply:
Does the "alternative considered" section help at all?

I'm not sure if Eager StructuredTensors can meet our performance requirements in certain operations (filtering / slicing / concatnating, etc)*, and is suitable for our development as we don't want to use TF to compute our custom statistics.

*Even if they can be easily implemented, given the fact that StructuredTensor is a collection of a bunch of EagerTensors (py objects) and nested metadata (still py objects), I'm not sure they can easily be implemented efficiently.